src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,315 @@
+                    i.RecentStats
+        ]
+        |> String.concat "\n"
+
+    static member AddInstrumentation(cacheId) =
+        if observedCaches.ContainsKey cacheId then
+            invalidArg "cacheId" $"cache with name {cacheId} already exists"
+
+        observedCaches[cacheId] <- new CacheMetrics(cacheId)
+
+    static member RemoveInstrumentation(cacheId) =
+        observedCaches[cacheId].Dispose()
+        observedCaches.TryRemove(cacheId) |> ignore
+
+type EntityPool<'Key, 'Value>(totalCapacity, cacheId) =
+    let pool = ConcurrentBag<CachedEntity<'Key, 'Value>>()
+    let mutable created = 0
+
+    let overCapacity =
+        CacheMetrics.Meter.CreateCounter<int64>("over-capacity", "count", cacheId)
+
+    member _.Acquire(key, value) =
+        match pool.TryTake() with
+        | true, entity -> entity.ReUse(key, value)
+        | _ ->
+            if Interlocked.Increment &created > totalCapacity then
+                overCapacity.Add 1L
+
+            CachedEntity(key, value).WithNode()
+
+    member _.Reclaim(entity: CachedEntity<'Key, 'Value>) =
+        if pool.Count < totalCapacity then
+            pool.Add(entity)
+
+module Cache =
+    // During testing a lot of compilations are started in app domains and subprocesses.
+    // This is a reliable way to pass the override to all of them.
+    [<Literal>]
+    let private overrideVariable = "FSHARP_CACHE_OVERRIDE"
+
+    /// Use for testing purposes to reduce memory consumption in testhost and its subprocesses.
+    let OverrideCapacityForTesting () =
+        Environment.SetEnvironmentVariable(overrideVariable, "true", EnvironmentVariableTarget.Process)
+
+    let applyOverride (capacity: int) =
+        match Environment.GetEnvironmentVariable(overrideVariable) with
+        | NonNull _ when capacity > 1024 -> 1024
+        | _ -> capacity
+
+[<Sealed; NoComparison; NoEquality>]
+[<DebuggerDisplay("{GetStats()}")>]
+type Cache<'Key, 'Value when 'Key: not null and 'Key: equality>
+    internal (totalCapacity, headroom, cts: CancellationTokenSource, ?name, ?observeMetrics) =
+
+    let instanceId = defaultArg name (Guid.NewGuid().ToString())
+
+    let observeMetrics = defaultArg observeMetrics false
+
+    do
+        if observeMetrics then
+            CacheMetrics.AddInstrumentation instanceId
+
+    let meter = CacheMetrics.Meter
+    let hits = meter.CreateCounter<int64>("hits", "count", instanceId)
+    let misses = meter.CreateCounter<int64>("misses", "count", instanceId)
+    let evictions = meter.CreateCounter<int64>("evictions", "count", instanceId)
+
+    let evictionFails =
+        meter.CreateCounter<int64>("eviction-fails", "count", instanceId)
+
+    let pool = EntityPool<'Key, 'Value>(totalCapacity, instanceId)
+
+    let store =
+        ConcurrentDictionary<'Key, CachedEntity<'Key, 'Value>>(Environment.ProcessorCount, totalCapacity)
+
+    let evictionQueue = LinkedList<CachedEntity<'Key, 'Value>>()
```

Comment:
ConcurrentDictionary stores the key, and CachedEntity stores the key again.

Am I right assuming we only ever need the key within CachedEntity for eviction?
If yes, what if CachedEntity  did not repeat the key, ConcurrentDictionary holds it.
And `evictionQueue` would be a queue of struct tuples (key,CachedEntity<value>).


Does that makes sense?
Are there other tradeoffs?

Reply:
Hm in the end everything is stored in both the `store` and both `evictionQueue` as same reference to objects from entity pool.

So reducing storage in CachedEntity and increasing storage for evictionQueue (by adding key) would not save anything, because the references are now already shared anyway.


[I will keep the comment here if anyone has the same idea for reducing memory footprint]